From 8a92cf3a7bf98b5c236956e33cd7da87c41fefcd Mon Sep 17 00:00:00 2001
From: Lu Jialin <lujialin4@huawei.com>
Date: Mon, 3 Jul 2023 15:45:33 +0800
Subject: [PATCH 611/780] [Huawei] PSI: Introduce pressure.stat in psi

Offering: HULK
HULK inclusion
category: feature
bugzilla: 188780

-------------------------------

Introduce pressure.stat in psi for cgroupv1 and system, which will show
the fine grained time tracking for cgroup memory reclaim.

for example:

/test # cat /tmp/cpuacct/test/pressure.stat
cgroup_memory_reclaim
some avg10=45.78 avg60=10.40 avg300=2.26 total=13491160
full avg10=0.00 avg60=0.00 avg300=0.00 total=0

Signed-off-by: Lu Jialin <lujialin4@huawei.com>
---
 include/linux/psi.h    |  1 +
 kernel/cgroup/cgroup.c | 17 ++++++++++++
 kernel/sched/psi.c     | 60 ++++++++++++++++++++++++++++++++++++++++++
 3 files changed, 78 insertions(+)

diff --git a/include/linux/psi.h b/include/linux/psi.h
index 41e01aa7508c..b0f6f2d0c225 100644
--- a/include/linux/psi.h
+++ b/include/linux/psi.h
@@ -32,6 +32,7 @@ int psi_show(struct seq_file *s, struct psi_group *group, enum psi_res res);
 
 #ifdef CONFIG_PSI_FINE_GRAINED
 void psi_memstall_type_change(unsigned int type);
+int psi_stat_show(struct seq_file *s, struct psi_group *group);
 #else
 static inline void psi_memstall_type_change(unsigned int type) {}
 #endif
diff --git a/kernel/cgroup/cgroup.c b/kernel/cgroup/cgroup.c
index 1bb3e00d3763..2227bdf7c828 100644
--- a/kernel/cgroup/cgroup.c
+++ b/kernel/cgroup/cgroup.c
@@ -3783,6 +3783,16 @@ static void cgroup_pressure_release(struct kernfs_open_file *of)
 	psi_trigger_destroy(ctx->psi.trigger);
 }
 
+#ifdef CONFIG_PSI_FINE_GRAINED
+static int cgroup_psi_stat_show(struct seq_file *seq, void *v)
+{
+	struct cgroup *cgrp = seq_css(seq)->cgroup;
+	struct psi_group *psi = cgroup_ino(cgrp) == 1 ? &psi_system : cgrp->psi;
+
+	return psi_stat_show(seq, psi);
+}
+#endif
+
 #ifdef CONFIG_PSI_CGROUP_V1
 struct cftype cgroup_v1_psi_files[] = {
 	{
@@ -3818,6 +3828,13 @@ struct cftype cgroup_v1_psi_files[] = {
 		.poll = cgroup_pressure_poll,
 		.release = cgroup_pressure_release,
 	},
+#endif
+#ifdef CONFIG_PSI_FINE_GRAINED
+	{
+		.name = "pressure.stat",
+		.flags = CFTYPE_NO_PREFIX,
+		.seq_show = cgroup_psi_stat_show,
+	},
 #endif
 	{ } /* terminate */
 };
diff --git a/kernel/sched/psi.c b/kernel/sched/psi.c
index 13dc3c253d44..ca78a03b86f2 100644
--- a/kernel/sched/psi.c
+++ b/kernel/sched/psi.c
@@ -1708,6 +1708,63 @@ static const struct proc_ops psi_cpu_proc_ops = {
 	.proc_release	= psi_fop_release,
 };
 
+#ifdef CONFIG_PSI_FINE_GRAINED
+static const char *const psi_stat_names[] = {
+	"cgroup_memory_reclaim",
+};
+
+int psi_stat_show(struct seq_file *m, struct psi_group *group)
+{
+	struct psi_group_ext *psi_ext;
+	unsigned long avg[3] = {0, };
+	int i, w;
+	bool is_full;
+	u64 now, total;
+
+	if (static_branch_likely(&psi_disabled))
+		return -EOPNOTSUPP;
+
+	psi_ext = to_psi_group_ext(group);
+	mutex_lock(&group->avgs_lock);
+	now = sched_clock();
+	collect_percpu_times(group, PSI_AVGS, NULL);
+	if (now >= group->avg_next_update)
+		group->avg_next_update = update_averages(group, now);
+	mutex_unlock(&group->avgs_lock);
+	for (i = 0; i < NR_PSI_STAT_STATES; i++) {
+		is_full = i % 2;
+		for (w = 0; w < 3; w++)
+			avg[w] = psi_ext->avg[i][w];
+		total = div_u64(psi_ext->total[PSI_AVGS][i], NSEC_PER_USEC);
+		if (!is_full)
+			seq_printf(m, "%s\n", psi_stat_names[i / 2]);
+		seq_printf(m, "%s avg10=%lu.%02lu avg60=%lu.%02lu avg300=%lu.%02lu total=%llu\n",
+			   is_full ? "full" : "some",
+			   LOAD_INT(avg[0]), LOAD_FRAC(avg[0]),
+			   LOAD_INT(avg[1]), LOAD_FRAC(avg[1]),
+			   LOAD_INT(avg[2]), LOAD_FRAC(avg[2]),
+			   total);
+	}
+	return 0;
+}
+static int system_psi_stat_show(struct seq_file *m, void *v)
+{
+	return psi_stat_show(m, &psi_system);
+}
+
+static int psi_stat_open(struct inode *inode, struct file *file)
+{
+	return single_open(file, system_psi_stat_show, NULL);
+}
+
+static const struct proc_ops psi_stat_proc_ops = {
+	.proc_open	= psi_stat_open,
+	.proc_read	= seq_read,
+	.proc_lseek	= seq_lseek,
+	.proc_release	= psi_fop_release,
+};
+#endif
+
 #ifdef CONFIG_IRQ_TIME_ACCOUNTING
 static int psi_irq_show(struct seq_file *m, void *v)
 {
@@ -1744,6 +1801,9 @@ static int __init psi_proc_init(void)
 		proc_create("pressure/cpu", 0, NULL, &psi_cpu_proc_ops);
 #ifdef CONFIG_IRQ_TIME_ACCOUNTING
 		proc_create("pressure/irq", 0, NULL, &psi_irq_proc_ops);
+#endif
+#ifdef CONFIG_PSI_FINE_GRAINED
+		proc_create("pressure/stat", 0, NULL, &psi_stat_proc_ops);
 #endif
 	}
 	return 0;
-- 
2.17.1

